File: labs/lab13_carry_save/carry_save.cpp
// Lab 13 – Carry Save Adder + CSA-based Multiplication
// CSA mode input:  A B C n
// Multiply mode:   run binary with --multiply flag (uses built-in demo)

#include <iostream>
#include <vector>
#include <string>
#include <cstring>
using namespace std;

pair<int,int> fullAdder(int a, int b, int cin) {
    return { a ^ b ^ cin, (a & b) | (b & cin) | (a & cin) };
}

// CSA: reduce A+B+C to (sum_vec, carry_vec)
// A+B+C == sum_vec + (carry_vec << 1)
pair<int,int> csaAdd(int A, int B, int C, int n) {
    int S = 0, K = 0;
    for (int i = 0; i < n; i++) {
        auto [s, k] = fullAdder((A>>i)&1, (B>>i)&1, (C>>i)&1);
        S |= s << i;
        K |= k << i;
    }
    return { S, K };
}

string toBinary(long long v, int n) {
    string s(n, '0');
    for (int i = 0; i < n; i++) s[n-1-i] = '0' + ((v>>i)&1);
    return s;
}

void printCSA(int A, int B, int C, int n) {
    auto [S, K] = csaAdd(A, B, C, n);
    int result  = S + (K << 1);

    cout << "\nCarry Save Addition (" << n << "-bit)\n";
    cout << "  A = " << toBinary(A,n) << "  (" << A << ")\n";
    cout << "  B = " << toBinary(B,n) << "  (" << B << ")\n";
    cout << "  C = " << toBinary(C,n) << "  (" << C << ")\n\n";
    cout << "  CSA Stage:\n";
    cout << "  Sum vector   S = " << toBinary(S,n) << "  (" << S << ")\n";
    cout << "  Carry vector K = " << toBinary(K,n) << "  (" << K << ")\n";
    cout << "  K shifted (K<<1) = " << toBinary((long long)K<<1, n+1)
         << "  (" << (K<<1) << ")\n\n";
    cout << "  Final add S + (K<<1) = " << result << "\n";
    cout << "  Expected A+B+C      = " << A+B+C << "\n";
    cout << "  Match: " << (result == A+B+C ? "OK" : "FAIL") << "\n";
}

long long csaMultiply(int M, int Q, int n) {
    vector<int> Mbits(n), Qbits(n);
    for (int i = 0; i < n; i++) { Mbits[i]=(M>>i)&1; Qbits[i]=(Q>>i)&1; }

    // generate n partial products (each shifted left by i)
    vector<long long> pp(n, 0);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            pp[i] |= (long long)(Mbits[j] & Qbits[i]) << (i+j);

    // CSA tree: reduce while >2 operands
    while (pp.size() > 2) {
        vector<long long> next;
        size_t i = 0;
        while (i + 2 < pp.size()) {
            auto [S, K] = csaAdd((int)pp[i], (int)pp[i+1], (int)pp[i+2], 2*n);
            next.push_back(S);
            next.push_back((long long)K << 1);
            i += 3;
        }
        while (i < pp.size()) next.push_back(pp[i++]);
        pp = next;
    }

    long long product = 0;
    for (auto v : pp) product += v;
    return product;
}

void printMultiply(int M, int Q, int n) {
    long long product  = csaMultiply(M, Q, n);
    long long expected = (long long)M * Q;
    cout << "\nCSA Multiplication (" << n << "-bit)\n";
    cout << "  M = " << toBinary(M,n) << "  (" << M << ")\n";
    cout << "  Q = " << toBinary(Q,n) << "  (" << Q << ")\n";
    cout << "  Product  = " << product << "\n";
    cout << "  Expected = " << expected << "\n";
    cout << "  Match: " << (product == expected ? "OK" : "FAIL") << "\n";
}

int main(int argc, char* argv[]) {
    bool multiplyMode = (argc >= 2 && strcmp(argv[1], "--multiply") == 0);

    if (multiplyMode && argc >= 4) {
        int M = atoi(argv[2]);
        int Q = atoi(argv[3]);
        int n = argc >= 5 ? atoi(argv[4]) : 4;
        printMultiply(M, Q, n);
        return 0;
    }

    int A, B, C, n;
    cout << "Enter A B C n (e.g. 13 11 7 5): ";
    cin >> A >> B >> C >> n;
    printCSA(A, B, C, n);
    return 0;
}

File: labs/lab13_carry_save/carry_save.py
#!/usr/bin/env python3
"""Lab 13 – Carry Save Adder (CSA) and Carry Save Multiplication.

A Carry Save Adder reduces three N-bit numbers to two N-bit numbers
(sum vector + carry vector) in O(1) time, deferring carry propagation.
This is used in array multipliers and other hardware to speed up
multi-operand addition.

Run:
  python3 carry_save.py                     # demo
  python3 carry_save.py <A> <B> <C>         # add three numbers
  python3 carry_save.py --multiply <M> <Q>  # CSA-based multiplication
"""

import sys


# ---------------------------------------------------------------------------
# Gate primitives
# ---------------------------------------------------------------------------

def full_adder_bits(a: int, b: int, cin: int) -> tuple[int, int]:
    return a ^ b ^ cin, (a & b) | (b & cin) | (a & cin)


# ---------------------------------------------------------------------------
# Carry Save Adder (single stage)
# ---------------------------------------------------------------------------

def carry_save_add(A: int, B: int, C: int, n: int) -> tuple[int, int]:
    """Add three n-bit integers using a Carry Save Adder.

    Returns (sum_vector, carry_vector) such that:
        A + B + C  =  sum_vector + (carry_vector << 1)

    The carry vector must be shifted left 1 bit before the final add.
    """
    S = 0  # sum bits
    K = 0  # carry bits

    for i in range(n):
        ai = (A >> i) & 1
        bi = (B >> i) & 1
        ci = (C >> i) & 1
        s, k = full_adder_bits(ai, bi, ci)
        S |= s << i
        K |= k << i

    return S, K


def csa_final_add(S: int, K: int) -> int:
    """Combine sum and carry vectors using a final ripple-carry add."""
    return S + (K << 1)


# ---------------------------------------------------------------------------
# Display helpers
# ---------------------------------------------------------------------------

def fmt(val: int, n: int) -> str:
    return format(val & ((1 << n) - 1), f"0{n}b")


def print_csa(A: int, B: int, C: int, n: int):
    S, K = carry_save_add(A, B, C, n)
    result = csa_final_add(S, K)

    print(f"\nCarry Save Addition  ({n}-bit)")
    print(f"  A = {fmt(A, n)}  ({A})")
    print(f"  B = {fmt(B, n)}  ({B})")
    print(f"  C = {fmt(C, n)}  ({C})")
    print()
    print(f"  CSA Stage:")
    print(f"  S (sum vector)   = {fmt(S, n)}  ({S})")
    print(f"  K (carry vector) = {fmt(K, n)}  ({K})")
    print(f"  K << 1           = {fmt(K<<1, n+1)}  ({K<<1})")
    print()
    print(f"  Final add: S + (K<<1) = {result}")
    print(f"  Expected : A+B+C      = {A+B+C}")
    print(f"  Match    : {'✓' if result == A+B+C else '✗'}")


# ---------------------------------------------------------------------------
# CSA-based Multiplication (Carry Save Array Multiplier)
# ---------------------------------------------------------------------------

def csa_multiply(M: int, Q: int, n: int):
    """Multiply two unsigned n-bit numbers using a CSA tree.

    Generates partial products and reduces them with CSA stages,
    finishing with a final ripple-carry add.

    Returns (product, trace) where trace lists each CSA reduction step.
    """
    M_bits = [(M >> i) & 1 for i in range(n)]
    Q_bits = [(Q >> i) & 1 for i in range(n)]

    # Generate partial products: PP[i] = M * Q_bits[i] (shifted left by i)
    partial_products = []
    for i in range(n):
        pp = 0
        for j in range(n):
            pp |= (M_bits[j] & Q_bits[i]) << j
        partial_products.append(pp << i)

    trace = [list(partial_products)]

    # Reduce using CSA tree until 2 operands remain
    operands = list(partial_products)
    while len(operands) > 2:
        new_operands = []
        i = 0
        while i + 2 < len(operands):
            S, K = carry_save_add(operands[i], operands[i+1], operands[i+2], 2*n)
            new_operands.append(S)
            new_operands.append(K << 1)
            i += 3
        while i < len(operands):
            new_operands.append(operands[i])
            i += 1
        operands = new_operands
        trace.append(list(operands))

    product = sum(operands)
    return product, trace


def print_csa_multiply(M: int, Q: int, n: int):
    product, trace = csa_multiply(M, Q, n)
    expected = M * Q

    print(f"\nCSA-based Multiplication  ({n}-bit)")
    print(f"  M = {fmt(M, n)}  ({M})")
    print(f"  Q = {fmt(Q, n)}  ({Q})")
    print()
    print(f"  Partial Products:")
    for i, pp in enumerate(trace[0]):
        print(f"    PP[{i}] = {fmt(pp, 2*n)}  ({pp})")

    if len(trace) > 1:
        print(f"\n  CSA Reduction Stages:")
        for s, stage in enumerate(trace[1:], 1):
            print(f"  Stage {s}:")
            for j, v in enumerate(stage):
                print(f"    [{j}] = {fmt(v, 2*n)}  ({v})")

    print(f"\n  Product  = {product}")
    print(f"  Expected = {expected}")
    print(f"  Match    : {'✓' if product == expected else '✗'}")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    if "--multiply" in sys.argv:
        idx = sys.argv.index("--multiply")
        if idx + 2 < len(sys.argv):
            M = int(sys.argv[idx + 1])
            Q = int(sys.argv[idx + 2])
            n = int(sys.argv[idx + 3]) if idx + 3 < len(sys.argv) else max(M, Q).bit_length()
            print_csa_multiply(M, Q, n)
        else:
            print("Usage: carry_save.py --multiply <M> <Q> [bits]")
        return

    args = [a for a in sys.argv[1:] if not a.startswith("--")]
    if len(args) >= 3:
        A, B, C = int(args[0]), int(args[1]), int(args[2])
        n = max(A, B, C).bit_length() + 1
        print_csa(A, B, C, n)
        return

    # Demo
    print("=" * 50)
    print("Carry Save Adder Demo")
    print("=" * 50)
    print_csa(13, 11, 7, 5)
    print_csa(15, 15, 15, 5)

    print("\n" + "=" * 50)
    print("CSA-based Multiplication Demo")
    print("=" * 50)
    print_csa_multiply(6, 5, 4)
    print_csa_multiply(13, 11, 4)
    print_csa_multiply(7, 7, 4)


if __name__ == "__main__":
    main()
